File: src/Crypto/SqliteWasmBlazor.Crypto/Abstractions/PoolNaming.cs
// SqliteWasmBlazor - Minimal EF Core compatible provider
// MIT License

namespace SqliteWasmBlazor;

/// <summary>
/// Names the SAH pool holds that are not databases an app opens. Today
/// that is one shape: the parked copy a validated import leaves behind
/// while it decides whether to keep what arrived.
/// </summary>
public static class PoolNaming
{
    /// <summary>
    /// Suffix of the entry an import parks the previous content under
    /// (<c>TodoDb.db</c> → <c>TodoDb.db.import-park</c>).
    ///
    /// <para>
    /// A validated import writes the incoming file under the database's
    /// real name — page AAD binds ciphertext to the database path, so a
    /// file written anywhere else would not decrypt once moved — and parks
    /// what was there under this suffix. Renames are metadata-only, so the
    /// parked bytes are untouched and go back exactly as they were if the
    /// validator refuses. On success the park is dropped.
    /// </para>
    /// <para>
    /// A park outlives the import only if the tab dies mid-flight; the next
    /// import sweeps whatever it finds. UI listing pool content should hide
    /// these entries — <see cref="IsImportPark"/> identifies them — since
    /// nothing opens them and they exist for seconds at a time.
    /// </para>
    /// </summary>
    public const string ImportParkSuffix = ".import-park";

    /// <summary>
    /// True when <paramref name="databaseName"/> is a parked import copy
    /// rather than a database of its own.
    /// </summary>
    /// <param name="databaseName">Pool entry name, e.g. <c>TodoDb.db</c>.</param>
    public static bool IsImportPark(string databaseName)
    {
        ArgumentNullException.ThrowIfNull(databaseName);
        return databaseName.EndsWith(ImportParkSuffix, StringComparison.Ordinal);
    }

    /// <summary>
    /// The park name for <paramref name="databaseName"/>.
    /// </summary>
    /// <param name="databaseName">Database whose content is being parked.</param>
    public static string ImportParkFor(string databaseName)
    {
        ArgumentException.ThrowIfNullOrWhiteSpace(databaseName);
        return $"{databaseName}{ImportParkSuffix}";
    }
}

File: src/Base/SqliteWasmBlazor/TypeScript-Common/src/pool-naming.ts
// pool-naming.ts
//
// The pool entry names that are not databases. Two shapes:
//
//   *.import-park   the previous content of a database a validated import
//                   is replacing. Mirrors PoolNaming.ImportParkSuffix on
//                   the C# side, which owns the park/restore bookkeeping —
//                   the worker only knows the shape so it can put a park
//                   back when the session that made it never got to.
//
//   *-tmp           a chunked write in progress (encrypt/decrypt in place,
//                   single- or multi-DB import commit). Promoted over the
//                   real name by atomicReplaceFile, so an entry under one
//                   of these names is by definition an unfinished write.
//
// Both outlive their session only when a tab dies mid-flight or the
// platform closes the pool's access handles underneath a commit; the
// worker's init sweep is where they are collected.

/** Mirrors <c>PoolNaming.ImportParkSuffix</c>. */
export const IMPORT_PARK_SUFFIX = '.import-park';

/** Suffix of the temp slot a chunked plain→encrypted transition writes. */
export const ENCRYPT_TMP_SUFFIX = '.encrypt-tmp';

/** Suffix of the temp slot a chunked encrypted→plain transition writes. */
export const DECRYPT_TMP_SUFFIX = '.decrypt-tmp';

/** Suffix of the temp slot a single-DB import commit writes. */
export const SINGLE_IMPORT_TMP_SUFFIX = '.single-import-tmp';

/** Suffix of the temp slot a multi-DB (.dbs) import commit writes. */
export const MULTI_IMPORT_TMP_SUFFIX = '.multi-import-tmp';

/** Suffix of the temp slot a whole-pool (.eds) import commit writes. */
export const POOL_IMPORT_TMP_SUFFIX = '.import-tmp';

const TEMP_SLOT_SUFFIXES = [
    ENCRYPT_TMP_SUFFIX,
    DECRYPT_TMP_SUFFIX,
    SINGLE_IMPORT_TMP_SUFFIX,
    MULTI_IMPORT_TMP_SUFFIX,
    POOL_IMPORT_TMP_SUFFIX,
] as const;

/** True when <paramref name="name"/> is a parked database copy. */
export function isImportPark(name: string): boolean {
    return name.endsWith(IMPORT_PARK_SUFFIX);
}

/** The park name for <paramref name="name"/>. */
export function importParkFor(name: string): string {
    return `${name}${IMPORT_PARK_SUFFIX}`;
}

/** The database name a park belongs to. Throws on a non-park name. */
export function databaseNameForPark(park: string): string {
    if (!isImportPark(park)) {
        throw new Error(`databaseNameForPark: '${park}' is not a park name`);
    }
    return park.slice(0, park.length - IMPORT_PARK_SUFFIX.length);
}

/** True when <paramref name="name"/> is an unfinished chunked write. */
export function isTempSlot(name: string): boolean {
    return TEMP_SLOT_SUFFIXES.some(suffix => name.endsWith(suffix));
}

/** What the init sweep does to one entry. */
export type PoolSweepAction =
    | { kind: 'restore'; park: string; database: string }
    | { kind: 'drop'; name: string };

/**
 * What a session that died mid-flight left behind, and what to do with it.
 *
 * A park whose database is absent is that database — the restore that
 * would have put it back never ran, and nothing else holds those bytes —
 * so it goes back under its own name. A park whose database is present is
 * left alone: the names cannot say whether it outlived a finished import
 * or a rollback that got half-way, and the next import's parking pass
 * settles it. Parks are never dropped here.
 *
 * A temp slot is dropped. Its content only ever becomes a database through
 * atomicReplaceFile, inside the one operation that wrote it; left in place
 * it lists as a database nothing can open.
 */
export function planPoolSweep(names: readonly string[]): PoolSweepAction[] {
    const present = new Set(names.filter(name => !isImportPark(name)));
    const actions: PoolSweepAction[] = [];
    for (const name of names) {
        if (isImportPark(name)) {
            const database = databaseNameForPark(name);
            if (!present.has(database)) {
                present.add(database);
                actions.push({kind: 'restore', park: name, database});
            }
            continue;
        }
        if (isTempSlot(name)) {
            actions.push({kind: 'drop', name});
        }
    }
    return actions;
}
